File: advent_of_code/2015/2/solution.py
import re
from functools import reduce

class Box():
    def __init__(self, dimensions_string):
        self.dimensions_string = dimensions_string
        self.parsed_dimensions = self.parse_dimensions(dimensions_string)
        self.l = self.parsed_dimensions[0]
        self.w = self.parsed_dimensions[1]
        self.h = self.parsed_dimensions[2]
        self.surfaces = self.calculate_surfaces()
        self.slack = self.calculate_slack()
        self.ribbon = self.calculate_ribbon()

    def __repr__(self):
        return f'_Box: {self.dimensions_string}_'

    def parse_dimensions(self, dimensions_string):
        dimensions = dimensions_string.split('x')
        return list(map(lambda n : int(n), dimensions))

    def calculate_surfaces(self):
        return [
                   2 * self.l * self.w,
                   2 * self.w * self.h,
                   2 * self.h * self.l
               ]

    def calculate_slack(self):
        surfaces_sorted = sorted(self.surfaces)
        return surfaces_sorted[0] / 2

    def total_paper_needed(self):
        return reduce(lambda a, b : a + b, self.surfaces) + self.slack

    def calculate_ribbon(self):
        sorted_dimensions = sorted(self.parsed_dimensions)
        ribbon = (sorted_dimensions[0] + sorted_dimensions[1]) * 2
        bow = self.h * self.l * self.w
        return ribbon + bow
        

class Solver():
    def __init__(self):
        self.input = open('input.txt', 'r').read()
        self.boxes = []
        self.total_paper_needed = 0
        self.total_ribbon_needed = 0

    def parse_input(self):
        dimensions_collection = re.findall(r'\d+x\d+x\d+', self.input)
        for dimensions in dimensions_collection:
            self.boxes.append(Box(dimensions))

    def calculate_total_paper_needed(self):
        for box in self.boxes:
            self.total_paper_needed += box.total_paper_needed()
        print(f'Total paper needed is {self.total_paper_needed}')

    def calculate_total_ribbon_needed(self):
        for box in self.boxes:
            self.total_ribbon_needed += box.calculate_ribbon()

    def solve(self):
        self.parse_input()
        self.calculate_total_paper_needed()
        self.calculate_total_ribbon_needed()
        print(f'Total ribbon needed is {self.total_ribbon_needed}')

solver = Solver()
solver.solve()

File: advent_of_code/2015/2/solution.js
/*
 * the input is in the format of
 * lxwxh
 * where l is length, w is width, and h is height.
 * surface area of the box is 2*l*w + 2*w*h + 2*h*l
 * every box needs an extra of the smallest side too
 * question: how many total sq ft of wrapping paper should they order?
 *
 * the input file is only 1k lines long, 7.9K big, thus no need to use streams
 *
 * after solving it I realised I could have used a re for the whole file -
 * it makes turning it into array, and clearning unnecessary - could have been
 * a series of matches.
 */

const fs = require("fs");

const getBoxSizes = () => {
  return new Promise((resolve, reject) => {
    const parseBoxSizesFile = (err, data) => {
      if (err) reject(err);
      resolve(data);
    };
    fs.readFile(`${__dirname}/input.txt`, "utf8", parseBoxSizesFile);
  });
};

/**
 * @param {String} boxSizeString
 * @returns {Object} containing three keys, l,h, and w, as ints
 */
const parseBoxSizeString = (boxSizeString) => {
  const sizeStringRe = /^(?<l>\d+)x(?<w>\d+)x(?<h>\d+)$/;
  const boxSizes = sizeStringRe.exec(boxSizeString).groups;
  for (let entry in boxSizes) {
    boxSizes[entry] = parseInt(boxSizes[entry], 10);
  }
  return boxSizes;
};

/**
 *
 * @param {Object} boxSizes
 * @returns {Number} amount of square feet needed for the wrapping
 */
const calculateAmountOfPaperNeeded = (boxSizeString) => {
  const boxSizes = parseBoxSizeString(boxSizeString);
  const { l, h, w } = boxSizes;
  const side1Surface = 2 * l * w;
  const side2Surface = 2 * h * w;
  const side3Surface = 2 * l * h;

  const surfaces = [side1Surface, side2Surface, side3Surface];
  surfaces.sort((a, b) => a - b);
  surfaces.push(surfaces[0] / 2); // add slack
  const totalPaperNeeded = surfaces.reduce(
    (acc, currentVal) => acc + currentVal,
    0
  );
  return totalPaperNeeded;
};

const calculateRibbonLength = (boxSizeString) => {
  const boxSizes = parseBoxSizeString(boxSizeString);
  const boxSizesAsArray = [];
  for (const size of Object.values(boxSizes)) {
    boxSizesAsArray.push(size);
  }
  let ribbonLength = 0;
  boxSizesAsArray.sort((a, b) => a - b);

  // the wrap-around
  for (let i = 0; i <= 1; i++) {
    ribbonLength += boxSizesAsArray[i] * 2;
  }

  // the bow
  let bowBit = boxSizesAsArray[0];
  for (let i = 1; i <= 2; i++) {
    bowBit *= boxSizesAsArray[i];
  }

  ribbonLength += bowBit;
  return ribbonLength;
};

const solve = async () => {
  const boxSizes = await getBoxSizes();
  let boxSizesArray = boxSizes.split("\n");
  boxSizesArray = boxSizesArray.filter((bs) => Boolean(bs));
  let totalFeetOfWrappingPaper = 0;
  let totalFeetOfRibbon = 0;
  for (let boxSize of boxSizesArray) {
    totalFeetOfWrappingPaper += calculateAmountOfPaperNeeded(boxSize);
    totalFeetOfRibbon += calculateRibbonLength(boxSize);
  }
  console.log(
    `The elves need ${totalFeetOfWrappingPaper} ft of wrapping paper, and ${totalFeetOfRibbon} ft of ribbon.`
  );
  return totalFeetOfWrappingPaper;
};

solve();
